File: 2021/python/1675.minimize-deviation-in-arr.py
def minimumDeviation(nums: List[int]) -> int:
        if not nums or len(nums) < 2:
            return 0
        # 1. turn all odd elements into even and enqueue to a max heap 
        # 2. find min and max value
        small = -nums[0] if nums[0] % 2 == 0 else -nums[0] * 2
        large = small
        pq = []
        for num in nums:
            if num % 2 == 1:
                num *= 2
            large = max(large, -num)
            small = min(small, -num)
            heapq.heappush(pq, -num)
        # 3. every time poll lartest element from pq, divide 2 and put it back to max heap 
        # 4. return until max element is odd number
        res = large - small
        while True:
            #print(heapq.nlargest(4, pq))
            top = heapq.heappop(pq)
            res = min(res, abs(top - large))
            if top % 2 == 1:
                return res
            top //= 2
            # maintain smallest element
            if top > large:
                large = top
            heapq.heappush(pq, top)
        return res

File: 2021/java/1675.minimize-deviation-in-arr.java
class Solution {
    public int minimumDeviation(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        // 1. all odd elements *2 become even,
        // 2. put all even elements into treeset, de-dup
        TreeSet<Integer> set = new TreeSet<>();
        for (int n : nums) {
            if (n % 2 == 1) n *= 2;
            set.add(n);
        }
        if (set.size() == 1) return 0;
        int res = Integer.MAX_VALUE;
        // 3. retrieve min and max element from set and compare, max / 2 and put back into set
        // 4. when set only have 1 element return 0 or lasgest element is odd, return res
        while (true) {
            if (set.size() == 1) return 0;
            int min = set.pollFirst();
            int max = set.pollLast();
            res = Math.min(res, max - min);
            if (max % 2 == 1) return res;
            set.add(min);
            set.add(max / 2);
        }
    }
}
